File: labs/24-sre-agent/go/main.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"regexp"
	"strings"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

var allowedCommands = []string{
	"systemctl restart",
	"kubectl rollout restart",
	"docker restart",
	"echo",
	"df",
}

func parseLogs(path string) []map[string]string {
	pattern := regexp.MustCompile(`^(\S+)\s+(\w+)\s+(.+)$`)
	b, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var entries []map[string]string
	for _, line := range strings.Split(string(b), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		match := pattern.FindStringSubmatch(line)
		if match != nil {
			entries = append(entries, map[string]string{"ts": match[1], "level": match[2], "message": match[3]})
		} else {
			entries = append(entries, map[string]string{"ts": "", "level": "UNKNOWN", "message": line})
		}
	}
	return entries
}

func summarizeErrors(entries []map[string]string) map[string]interface{} {
	var errorMessages []string
	counts := map[string]int{}
	for _, e := range entries {
		counts[e["level"]]++
		if e["level"] == "ERROR" || e["level"] == "FATAL" {
			errorMessages = append(errorMessages, e["message"])
		}
	}
	return map[string]interface{}{"error_messages": errorMessages, "counts": counts}
}

func diagnose(entries []map[string]string, client *openaiclient.Client) map[string]interface{} {
	if client == nil {
		return map[string]interface{}{
			"diagnosis": "Database appears unreachable (multiple connection timeouts and 503 errors). Disk usage is also elevated.",
			"commands": []string{
				"echo 'Checking database service status...'",
				"systemctl restart postgresql",
				"echo 'Monitoring disk usage...'",
				"df -h",
			},
			"risk": "medium",
		}
	}

	var lines []string
	for _, e := range entries {
		lines = append(lines, fmt.Sprintf("%s: %s", e["level"], e["message"]))
	}
	logText := strings.Join(lines, "\n")
	prompt := fmt.Sprintf(`You are an SRE agent. Diagnose the following logs and respond with JSON in this exact shape:
{
  "diagnosis": "short diagnosis",
  "commands": ["command1", "command2"],
  "risk": "low|medium|high"
}

Logs:
%s`, logText)

	resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
		Messages:       []openaiclient.Message{{Role: "user", Content: prompt}},
		Temperature:    floatPtr(0.2),
		MaxTokens:      intPtr(200),
		ResponseFormat: map[string]string{"type": "json_object"},
	})
	if err != nil {
		return map[string]interface{}{"diagnosis": fmt.Sprintf("Diagnosis failed: %v", err), "commands": []string{}, "risk": "unknown"}
	}
	content := resp.Choices[0].Message.Content
	var parsed map[string]interface{}
	if err := json.Unmarshal([]byte(content), &parsed); err != nil {
		return map[string]interface{}{"diagnosis": content, "commands": []string{}, "risk": "unknown"}
	}
	return parsed
}

func isCommandAllowed(command string) bool {
	for _, prefix := range allowedCommands {
		if strings.HasPrefix(strings.TrimSpace(command), prefix) {
			return true
		}
	}
	return false
}

func executeCommands(commands []string) []map[string]interface{} {
	var results []map[string]interface{}
	for _, cmd := range commands {
		fmt.Printf("  $ %s\n", cmd)
		if !isCommandAllowed(cmd) {
			fmt.Println("    -> SKIPPED (not in allowlist)")
			results = append(results, map[string]interface{}{"command": cmd, "status": "skipped", "output": ""})
			continue
		}
		result := exec.Command("sh", "-c", cmd)
		out, err := result.CombinedOutput()
		output := strings.TrimSpace(string(out))
		if output == "" {
			output = "<no output>"
		}
		if err != nil {
			fmt.Printf("    -> ERROR: %v\n", err)
			results = append(results, map[string]interface{}{"command": cmd, "status": "error", "output": err.Error()})
			continue
		}
		fmt.Printf("    -> exit 0\n")
		results = append(results, map[string]interface{}{"command": cmd, "status": "executed", "output": output})
	}
	return results
}

func floatPtr(f float64) *float64 {
	return &f
}

func intPtr(i int) *int {
	return &i
}

func main() {
	cfg := openaiclient.LoadConfig()
	var client *openaiclient.Client
	if cfg.APIKey != "" || strings.HasPrefix(cfg.BaseURL, "http://localhost") {
		client = openaiclient.NewClient()
	} else {
		log.Println("LLM client disabled: OPENAI_API_KEY not set")
	}

	logfile := "sample_app.log"
	if len(os.Args) > 1 {
		logfile = os.Args[1]
	}
	if _, err := os.Stat(logfile); err != nil {
		log.Fatalf("Log file not found: %s", logfile)
	}

	entries := parseLogs(logfile)
	fmt.Printf("Read %d log entries from %s\n", len(entries), logfile)
	diagnosis := diagnose(entries, client)

	fmt.Println("\n=== Diagnosis ===")
	fmt.Println(diagnosis["diagnosis"])
	fmt.Printf("Risk: %v\n", diagnosis["risk"])

	commands := []string{}
	if c, ok := diagnosis["commands"].([]interface{}); ok {
		for _, cmd := range c {
			commands = append(commands, cmd.(string))
		}
	}
	fmt.Println("\n=== Proposed remediation commands ===")
	for _, cmd := range commands {
		fmt.Printf("  - %s\n", cmd)
	}
	if len(commands) == 0 {
		fmt.Println("No remediation commands proposed.")
		return
	}

	reader := bufio.NewReader(os.Stdin)
	fmt.Print("\nExecute proposed commands? [y/n]: ")
	answer, _ := reader.ReadString('\n')
	answer = strings.TrimSpace(strings.ToLower(answer))
	if answer != "y" && answer != "yes" {
		fmt.Println("Execution cancelled by operator.")
		return
	}

	fmt.Println("\nExecuting commands...")
	results := executeCommands(commands)
	fmt.Println("\nExecution summary:")
	b, _ := json.MarshalIndent(results, "", "  ")
	fmt.Println(string(b))
}

File: labs/24-sre-agent/java/Main.java
import com.agentlabs.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {

    private static final Set<String> ALLOWED_COMMANDS = Set.of(
            "systemctl restart",
            "kubectl rollout restart",
            "docker restart",
            "echo",
            "df"
    );

    public static void main(String[] args) throws Exception {
        OpenAiClient client = null;
        try {
            client = new OpenAiClient();
        } catch (IllegalStateException exc) {
            System.out.println("LLM client disabled: " + exc.getMessage());
        }

        Path logPath = args.length > 0 ? Paths.get(args[0]) : Paths.get("..", "python", "sample_app.log");
        if (!Files.exists(logPath)) {
            System.err.println("Log file not found: " + logPath);
            System.exit(1);
        }

        List<Map<String, String>> entries = parseLogs(logPath);
        System.out.println("Read " + entries.size() + " log entries from " + logPath);

        Map<String, Object> diagnosis = diagnose(entries, client);
        System.out.println("\n=== Diagnosis ===");
        System.out.println(diagnosis.get("diagnosis"));
        System.out.println("Risk: " + diagnosis.getOrDefault("risk", "unknown"));

        @SuppressWarnings("unchecked")
        List<String> commands = (List<String>) diagnosis.getOrDefault("commands", List.of());
        System.out.println("\n=== Proposed remediation commands ===");
        if (commands.isEmpty()) {
            System.out.println("No remediation commands proposed.");
            return;
        }
        for (String cmd : commands) {
            System.out.println("  - " + cmd);
        }

        Scanner scanner = new Scanner(System.in);
        System.out.print("\nExecute proposed commands? [y/n]: ");
        String answer = scanner.nextLine().strip().toLowerCase();
        if (!answer.equals("y") && !answer.equals("yes")) {
            System.out.println("Execution cancelled by operator.");
            return;
        }

        System.out.println("\nExecuting commands...");
        List<Map<String, Object>> results = executeCommands(commands);
        System.out.println("\nExecution summary:");
        System.out.println(new ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(results));
    }

    private static List<Map<String, String>> parseLogs(Path path) throws IOException {
        Pattern pattern = Pattern.compile("^(\\S+)\\s+(\\w+)\\s+(.+)$");
        List<Map<String, String>> entries = new ArrayList<>();
        for (String line : Files.readAllLines(path)) {
            Matcher matcher = pattern.matcher(line);
            if (matcher.matches()) {
                Map<String, String> entry = new LinkedHashMap<>();
                entry.put("ts", matcher.group(1));
                entry.put("level", matcher.group(2));
                entry.put("message", matcher.group(3));
                entries.add(entry);
            } else {
                Map<String, String> entry = new LinkedHashMap<>();
                entry.put("ts", "");
                entry.put("level", "UNKNOWN");
                entry.put("message", line);
                entries.add(entry);
            }
        }
        return entries;
    }

    private static Map<String, Object> summarizeErrors(List<Map<String, String>> entries) {
        List<String> errorMessages = new ArrayList<>();
        Map<String, Integer> counts = new HashMap<>();
        for (Map<String, String> e : entries) {
            String level = e.get("level");
            counts.put(level, counts.getOrDefault(level, 0) + 1);
            if ("ERROR".equals(level) || "FATAL".equals(level)) {
                errorMessages.add(e.get("message"));
            }
        }
        return Map.of("error_messages", errorMessages, "counts", counts);
    }

    private static Map<String, Object> diagnose(List<Map<String, String>> entries, OpenAiClient client) throws Exception {
        if (client == null) {
            return Map.of(
                    "diagnosis", "Database appears unreachable (multiple connection timeouts and 503 errors). Disk usage is also elevated.",
                    "commands", List.of(
                            "echo 'Checking database service status...'",
                            "systemctl restart postgresql",
                            "echo 'Monitoring disk usage...'",
                            "df -h"
                    ),
                    "risk", "medium"
            );
        }
        StringBuilder logText = new StringBuilder();
        for (Map<String, String> e : entries) {
            logText.append(e.get("level")).append(": ").append(e.get("message")).append("\n");
        }
        String prompt = "You are an SRE agent. Diagnose the following logs and respond with JSON in this exact shape:\n{\n  \"diagnosis\": \"short diagnosis\",\n  \"commands\": [\"command1\", \"command2\"],\n  \"risk\": \"low|medium|high\"\n}\n\nLogs:\n" + logText;

        ObjectMapper mapper = new ObjectMapper();
        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();
        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", prompt);
        payload.set("messages", messages);
        payload.put("temperature", 0.2);
        payload.put("max_tokens", 200);
        ObjectNode responseFormat = mapper.createObjectNode();
        responseFormat.put("type", "json_object");
        payload.set("response_format", responseFormat);

        JsonNode response = client.chatCompletion(payload);
        String content = client.extractMessage(response).get("content").asText();
        try {
            return mapper.readValue(content, Map.class);
        } catch (Exception exc) {
            return Map.of("diagnosis", content, "commands", List.of(), "risk", "unknown");
        }
    }

    private static boolean isCommandAllowed(String command) {
        for (String prefix : ALLOWED_COMMANDS) {
            if (command.strip().startsWith(prefix)) {
                return true;
            }
        }
        return false;
    }

    private static List<Map<String, Object>> executeCommands(List<String> commands) {
        List<Map<String, Object>> results = new ArrayList<>();
        for (String cmd : commands) {
            System.out.println("  $ " + cmd);
            if (!isCommandAllowed(cmd)) {
                System.out.println("    -> SKIPPED (not in allowlist)");
                results.add(Map.of("command", cmd, "status", "skipped", "output", ""));
                continue;
            }
            try {
                ProcessBuilder pb = new ProcessBuilder("sh", "-c", cmd);
                pb.redirectErrorStream(true);
                Process process = pb.start();
                if (!process.waitFor(10, TimeUnit.SECONDS)) {
                    process.destroyForcibly();
                    results.add(Map.of("command", cmd, "status", "error", "output", "timed out"));
                    System.out.println("    -> ERROR: timed out");
                    continue;
                }
                String output = new String(process.getInputStream().readAllBytes()).strip();
                if (output.isEmpty()) {
                    output = "<no output>";
                }
                System.out.println("    -> exit " + process.exitValue());
                results.add(Map.of("command", cmd, "status", "executed", "output", output));
            } catch (Exception exc) {
                System.out.println("    -> ERROR: " + exc.getMessage());
                results.add(Map.of("command", cmd, "status", "error", "output", exc.getMessage()));
            }
        }
        return results;
    }
}
